### PR TITLE
feat: add vector index client

### PR DESCRIPTION
**Pull request by malandis**
Adds the new Vector Index Client, responses, and integration tests. Integration tests work against a local deployment of the service.

As such customers will not be able to use the client yet.

### CODE REVIEW

File: src/momento/internal/synchronous/_scs_grpc_manager.py
Comment (malandis) [inline src/momento/internal/synchronous/_scs_grpc_manager.py:24]: Will revert this once we cut over to the live service.

File: src/momento/internal/aio/_scs_grpc_manager.py
Comment (malandis) [inline src/momento/internal/aio/_scs_grpc_manager.py:29]: Will revert this once we cut over to the live service.

File: src/momento/vector_index_client_async.py
Comment (malandis) [inline src/momento/vector_index_client_async.py:41]: 🎗️ rename to `PreviewVectorIndexClientAsync`

File: tests/conftest.py
Comment (malandis) [inline tests/conftest.py:63]: 🎗️ Will remove this once we cut over to the live service.

**Review by pgautier404 — DISMISSED** [review]
LGTM, and I was able to run the tests against gRPC mock server.

File: src/momento/vector_index_client_async.py
Comment (cprice404) [inline src/momento/vector_index_client_async.py:27]: is this actually still a thing?

File: src/momento/cache_client.py
Comment (cprice404) [inline src/momento/cache_client.py:40]: what's going on with these changes?

**Comment by cprice404** [comment]
i kind of rushed through this, it's a big PR and I haven't had as much time to look at it as I need, but I didn't want to delay on getting some initial thoughts shared out

File: src/momento/internal/aio/_scs_control_client.py
Comment (malandis) [reply src/momento/internal/aio/_scs_control_client.py:41]: Yes that's right

File: src/momento/config/transport/grpc_configuration.py
Comment (cprice404) [reply src/momento/config/transport/grpc_configuration.py:23]: i don't really love adding more surface area to the configuration object APIs without a customer-facing use case. also isn't your local setup going to become obsolete as soon as the real service goes live, since it's not the same tech stack?

File: src/momento/internal/aio/_scs_grpc_manager.py
Comment (malandis) [reply src/momento/internal/aio/_scs_grpc_manager.py:37]: I intended to leave the grpc configuration stuff in to make local testing easier. I'm going to remove the hard coded provider info though.

File: src/momento/internal/aio/_vector_index_client.py
Comment (malandis) [reply src/momento/internal/aio/_vector_index_client.py:25]: I can do that. Here I was just paralleling the the name of the service in the protos.

File: src/momento/config/transport/grpc_configuration.py
Comment (cprice404) [reply src/momento/config/transport/grpc_configuration.py:23]: I won't block the PR on it if you feel strongly about it, it's not an unreasonable thing to support; I'm just doing my usual advocacy for keeping API footprint as small as possible

File: src/momento/cache_client.py
Comment (malandis) [reply src/momento/cache_client.py:40]: These hadn't been added to `momento.responses` for some reason, hence the really long `from momento.responses.data....`. While I was here I added those to `momento.responses` and moved these import lines below.

Comment (cprice404) [reply src/momento/cache_client.py:40]: :ack:

File: src/momento/config/vector_index_configurations.py
Comment (malandis) [reply src/momento/config/vector_index_configurations.py:32]: Thanks good catch!

File: src/momento/internal/aio/_vector_index_client.py
Comment (malandis) [reply src/momento/internal/aio/_vector_index_client.py]: refactored in [a06c8c1](https://github.com/momentohq/client-sdk-python/pull/358/commits/a06c8c1f392d8575c395221262e5c3ee05759d74); agree for the sake of clarity it makes sense

File: src/momento/config/vector_index_configurations.py
Comment (malandis) [reply src/momento/config/vector_index_configurations.py]: Had to do some refactoring to accomodate this. Remember the cache control service hosts the vector index control RPCs. Thus to have a dedicated vector index configuration, I needed to create a separate vector index control client and control grpc manager. In [6bb890e](https://github.com/momentohq/client-sdk-python/pull/358/commits/6bb890ee6a718f8cfc8586f018930efdbbccdb94) I created modules to house each of these and refactored the code.

In [e82cd9f](https://github.com/momentohq/client-sdk-python/pull/358/commits/e82cd9fe7d096eb1f28d93c74fa36d1c919dfeca) I introduced the vector index configuration objects. These have a transport strategy and client timeout but not a retry strategy.

File: src/momento/config/transport/grpc_configuration.py
Comment (cprice404) [reply src/momento/config/transport/grpc_configuration.py:23]: @malandis did you have any more thoughts on this in response to comments above?

File: src/momento/config/vector_index_configuration.py
Comment (cprice404) [inline src/momento/config/vector_index_configuration.py:23]: thanks for doing this, I think this is a more future-proof path.

File: src/momento/internal/synchronous/_vector_index_data_client.py
Comment (cprice404) [inline src/momento/internal/synchronous/_vector_index_data_client.py:67]: I think it's probably okay to change this in a subsequent PR if we need to, but we might need a wrapper type for this `list[float]` here, right? for forward-compat with numpy etc.?

File: src/momento/vector_index_client_async.py
Comment (cprice404) [reply src/momento/vector_index_client_async.py:29]: I don't know if we need these rosetta workarounds anymore, I feel like this got fixed in gRPC. Can address in a subsequent PR but I would love to get rid of this.

**Review by cprice404 — APPROVED** [review]
i've left a few comments on here that I'm not sure we've bottomed out on. the biggest one is the gRPC transport changes involving the certs, because that is a public API whose scope extends beyond this new preview client, so if we add that we can't really change our mind about it.

nothing in my comments that I consider to be an absolute blocker but just make sure you feel really confident about that decision before we merge.

File: src/momento/vector_index_client_async.py
Comment (malandis) [reply src/momento/vector_index_client_async.py:29]: I don't know if it's still a thing. I copied it from the data client. Will be good to revisit and can do in a future PR.

File: src/momento/config/transport/transport_strategy.py
Comment (pratik151192) [inline src/momento/config/transport/transport_strategy.py:51]: Would it make sense to have a separate gRPC configuration altogether for MVI as these don't seem related to cache? The only reason I'm saying is both these requirements may have divergent configurations over time, and I'm not sure you'd want to model all of them as `Optional` here because we don't want to break compatibility with the client configurations.

**Review by cprice404 — COMMENTED** [review]
Do we have a handle on what the Red CI is about here?

I've chatted with Michael enough about this one to where I am basically ready to ship it once CI is green and he gives a 👍 .

My only real concerns are:
1. Distinct Configuration object for Vector rather than re-using Cache
2. New namespace for Vector responses, and a ticket created to do the same kind of namespacing (via aliases) for the existing response types
3. would prefer not to add new cert APIs to config for now if we don't need them, but it's not a blocker
4. naming conventions and other API considerations are reasonably aligned with Pete's JS PR. (there may be changes required in his PR to accommodate that, but I want to make sure they are aligned.)

**Comment by malandis** [comment]
> Do we have a handle on what the Red CI is about here?
> 
> I've chatted with Michael enough about this one to where I am basically ready to ship it once CI is green and he gives a 👍 .
> 
> My only real concerns are:
> 
> 1. Distinct Configuration object for Vector rather than re-using Cache
We separate the config objects because the cache config had retries, and which aren't relevant to the vector client for now.  As for defaults, the vector add APIs have different latency requirements.

> 2. New namespace for Vector responses, and a ticket created to do the same kind of namespacing (via aliases) for the existing response types
> 3. would prefer not to add new cert APIs to config for now if we don't need them, but it's not a blocker
I thought we removed these? Let's double check for vestiges.

> 4. naming conventions and other API considerations are reasonably aligned with Pete's JS PR. (there may be changes required in his PR to accommodate that, but I want to make sure they are aligned.)
I left a common on his PR but we're otherwise aligned.

File: tests/utils.py
Comment (malandis) [inline tests/utils.py:13]:
```suggestion
    return f"python-test-{uuid_str()}"
```

We can remove that now :)

**Review by malandis — COMMENTED** [review]
One small change to go

**Review by cprice404 — DISMISSED** [review]

**Review by cprice404 — APPROVED** [review]

**Pull request merged**
